### PR TITLE
feat(active-directory): create @verdaccio/active-directory plugin

### PR DESCRIPTION
**Pull request by sergiohgz**
**Type:** feat
**Scope:** active-directory

The following has been addressed in the PR:

*  There is a related issue? **Part of #24**
*  Unit or Functional tests are included in the PR **Yes**

**Description:** create _@verdaccio/active-directory_ plugin and tests in Typescript
Note: there are some `// @ts-ignore` in code because the definition for `@types/activedirectory2` are not completed. In a future this will be removed when fixed.

### CODE REVIEW

File: plugins/active-directory/src/active-directory.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    connection.authenticate(username, password, (err, isAuthenticated): void => {
+      if (err) {
+        this.logger.warn(`AD - Active Directory authentication failed with error: ${err}`);
+        return cb(err);
+      }
+
+      if (!isAuthenticated) {
+        this.logger.warn(NotAuthMessage);
+        return cb(new Error(NotAuthMessage));
```
Comment (juanpicado) [inline plugins/active-directory/src/active-directory.ts]: Here we need to use the VerdaccioError or ErrorCode from commons-api

File: plugins/active-directory/tests/__mocks__/Logger.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { Logger } from '@verdaccio/types';
+
+const logger: Logger = {
+  warn: jest.fn(),
+  error: jest.fn(),
+  // fatal: jest.fn(),
```
Comment (juanpicado) [inline plugins/active-directory/tests/__mocks__/Logger.ts]: I would recommend this, multiple times I had to add it myself.

File: plugins/active-directory/src/active-directory.ts (hunk 1 above)
Comment (sergiohgz) [reply plugins/active-directory/src/active-directory.ts]: Would it looks like this?
```typescript
return cb(getUnauthorized(NotAuthMessage));
```

File: plugins/active-directory/tests/__mocks__/Logger.ts (hunk 2 above)
Comment (sergiohgz) [reply plugins/active-directory/tests/__mocks__/Logger.ts]: Do you mean uncomment this line? If that, it's done :wink:

File: plugins/active-directory/src/active-directory.ts (hunk 1 above)
Comment (sergiohgz) [reply plugins/active-directory/src/active-directory.ts]: Added types for all errors @juanpicado

**Comment by sergiohgz** [comment]
I created a PR to [DefinetlyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37702) to add the types needed to remove `// @ts-ignore`, following the definition shown in [activedirectory2](https://github.com/jsumners/node-activedirectory#documentation) documentation.

I'll create an issue from this comment to follow up and update when available

**Pull request merged**